client/ayon_core/hosts/substancepainter/api/lib.py

```diff
@@ -640,3 +642,58 @@ def _setup_prompt():
         return
 
     return project_mesh
+
+
+def get_export_presets_by_filtering(export_preset_name, channel_type_list):
+    new_maps = []
+
+    export_presets = get_export_presets()
+    resource_presets = substance_painter.export.list_resource_export_presets()
+    preset = next((preset for preset in resource_presets
+                   if preset.resource_id.name == (
+                       export_presets[export_preset_name])), None)
+
+    if preset is not None:
+        maps = preset.list_output_maps()
+        for channel_map in maps:
+            for n in channel_type_list:
+                if n in channel_map["fileName"]:
```

Comment:
Are we sure the channel type name is required to be in the "fileName". How reliable is this check?

Reply:
Ah no - if this is the output filename template then unfortunately no - that `_BaseColor` in `$mesh_$textureSet_BaseColor(_$colorSpace)(.$udim)` is the name set in the export template - but can also be `_CLR` or `_DIFF` or whatever a studio might set up in their templates.

This seems to be exactly what I struggled with.

Reply:
We can mention this limitation in the additional info. I am thinking we can make use of the project settings to deal with the specific channel names. Instead of the hard-coded list of layer_stack_channel_enum, we can allow users to add their own channel name with some preset names

Reply:
Yup - supporting custom ones could then be possible. Still dangerous - and preferably the Substance Painter API would allow this access out of the box to the source channels, etc

I wonder if that **channels** -> **srcMapName** that returns `basecolor` is exactly that?

Reply:
> Yup - supporting custom ones could then be possible. Still dangerous - and preferably the Substance Painter API would allow this access out of the box to the source channels, etc
> 
> I wonder if that **channels** -> **srcMapName** that returns `basecolor` is exactly that?

it is. I tried with my own custom export preset. with the fileName with CTL extension. 
```
[Python] output_map: [{'channels': [{'destChannel': 'R', 'srcChannel': 'R', 'srcMapName': 'basecolor', 'srcMapType': 'documentMap', 'srcPath': ''}, {'destChannel': 'G', 'srcChannel': 'G', 'srcMapName': 'basecolor', 'srcMapType': 'documentMap', 'srcPath': ''}, {'destChannel': 'B', 'srcChannel': 'B', 'srcMapName': 'basecolor', 'srcMapType': 'documentMap', 'srcPath': ''}, {'destChannel': 'A', 'srcChannel': 'L', 'srcMapName': 'opacity', 'srcMapType': 'documentMap', 'srcPath': ''}], 'fileName': '$mesh_$textureSet_CTL(_$colorSpace)(.$udim)', 'parameters': {'bitDepth': '8', 'dithering': False, 'fileFormat': 'png'}}, {'channels': [{'destChannel': 'L', 'srcChannel': 'L', 'srcMapName': 'roughness', 'srcMapType': 'documentMap', 'srcPath': ''}], 'fileName': '$mesh_$textureSet_Roughness(_$colorSpace)(.$udim)', 'parameters': {'bitDepth': '8', 'dithering': False, 'fileFormat': 'png'}}, {'channels': [{'destChannel': 'L', 'srcChannel': 'L', 'srcMapName': 'metallic', 'srcMapType': 'documentMap', 'srcPath': ''}], 'fileName': '$mesh_$textureSet_Metallic(_$colorSpace)(.$udim)', 'parameters': {'bitDepth': '8', 'dithering': False, 'fileFormat': 'png'}}, {'channels': [{'destChannel': 'R', 'srcChannel': 'R', 'srcMapName': 'Normal_DirectX', 'srcMapType': 'virtualMap', 'srcPath': ''}, {'destChannel': 'G', 'srcChannel': 'G', 'srcMapName': 'Normal_DirectX', 'srcMapType': 'virtualMap', 'srcPath': ''}, {'destChannel': 'B', 'srcChannel': 'B', 'srcMapName': 'Normal_DirectX', 'srcMapType': 'virtualMap', 'srcPath': ''}], 'fileName': '$mesh_$textureSet_Normal(_$colorSpace)(.$udim)', 'parameters': {'bitDepth': '16', 'dithering': False, 'fileFormat': 'png'}}, {'channels': [{'destChannel': 'L', 'srcChannel': 'L', 'srcMapName': 'height', 'srcMapType': 'documentMap', 'srcPath': ''}], 'fileName': '$mesh_$textureSet_Height(_$colorSpace)(.$udim)', 'parameters': {'bitDepth': '16', 'dithering': False, 'fileFormat': 'png'}}, {'channels': [{'destChannel': 'R', 'srcChannel': 'R', 'srcMapName': 'emissive', 'srcMapType': 'documentMap', 'srcPath': ''}, {'destChannel': 'G', 'srcChannel': 'G', 'srcMapName': 'emissive', 'srcMapType': 'documentMap', 'srcPath': ''}, {'destChannel': 'B', 'srcChannel': 'B', 'srcMapName': 'emissive', 'srcMapType': 'documentMap', 'srcPath': ''}], 'fileName': '$mesh_$textureSet_Emissive(_$colorSpace)(.$udim)', 'parameters': {'bitDepth': '8', 'dithering': False, 'fileFormat': 'png'}}]
```

Reply:
I tried to access the `srcMapName values but SP doesn't allow me to do it and gives weird value error. It just tells me that this is some kind of API limitations.
```
DEBUG: Exporting using preset: resource://your_assets/PBR Metallic Roughness_Kayla
DEBUG: http://localhost:5000 "POST /graphql HTTP/1.1" 200 686
DEBUG: Response <RestApiResponse [200]>
DEBUG: http://localhost:5000 "POST /graphql HTTP/1.1" 200 609
DEBUG: Response <RestApiResponse [200]>
Traceback (most recent call last):
  File "C:\Users\Kayla\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "D:\ayon-core\client\ayon_core\hosts\substancepainter\plugins\publish\collect_textureset_images.py", line 43, in process
    maps = get_parsed_export_maps(config)
  File "D:\ayon-core\client\ayon_core\hosts\substancepainter\api\lib.py", line 402, in get_parsed_export_maps
    outputs = substance_painter.export.list_project_textures(config)
  File "C:\Program Files/Adobe/Adobe Substance 3D Painter/resources/python/modules\substance_painter\export.py", line 69, in list_project_textures
    return _substance_painter.export.list_project_textures(json.dumps(json_config))
ValueError: For exportList item #0, the preset "resource://your_assets/PBR Metallic Roughness_Kayla" is undefined
```
The following code I have used which tried to parse the data from the config.
```python
    if preset is None:
        return {}
    maps = preset.list_output_maps()
    for channel_map in maps:
        for n in channel_type_list:
            if not channel_map.get("fileName"):
                continue
            for channel in channel_map['channels']:
                if channel.get('srcMapValue') == n.lower():
                    new_maps.append(channel)
```